File: epi_judge_python/convert_base.py
from test_framework import generic_test
import string


def convert_base(num_as_string: str, b1: int, b2: int) -> str:
    n = 0
    num_as_string, negative = (
        (num_as_string[1:].lower(), True)
        if num_as_string[0] == "-"
        else (num_as_string.lower(), False)
    )
    for i in range(len(num_as_string)):
        n += string.hexdigits.index(num_as_string[~i]) * (b1**i)
    res = "" if n != 0 else "0"
    while n > 0:
        res = string.hexdigits[n % b2].upper() + res
        n //= b2
    if negative:
        res = "-" + res
    return res


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "convert_base.py", "convert_base.tsv", convert_base
        )
    )

File: epi_judge_cpp/convert_base.cc
#include <string>
#include <unordered_map>

#include "test_framework/generic_test.h"
using std::string;

string ConvertBase(const string &num_as_string, int b1, int b2) {
  if (num_as_string.empty()) {
    return "";
  }
  unsigned long num = 0;
  for (auto it = num_as_string.begin(); it != num_as_string.end(); it++) {
    if (*it != '-') {
      num *= b1;
      num += (isdigit(*it) ? *it - '0' : *it - 'A' + 10);
    }
  }
  std::deque<char> nstr;
  if (num == 0) {
    return "0";
  }
  while (num) {
    int d = num % b2;
    nstr.push_front(d > 9 ? d - 10 + 'A' : '0' + d);
    num /= b2;
  }
  if (num_as_string.front() == '-') {
    nstr.push_front('-');
  }
  return string(nstr.begin(), nstr.end());
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"num_as_string", "b1", "b2"};
  return GenericTestMain(args, "convert_base.cc", "convert_base.tsv",
                         &ConvertBase, DefaultComparator{}, param_names);
}
